[InputGen] introduce variable space

manuelcandales opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #116138
* #116137
* #116136
* #116135
* #115698
* #115697
* #115696
* #115695
* #115694
* #115693
* #115692
* #115691
* #115605
* #115604
* __->__ #115603
* #115602
* #115601

This diff introduces the VariableSpace class.

A VariableSpace needs to be initialized with a variable type. Its purpose is to describe a space of values for that variable type. This space can be discrete (like a Python set) or it can be a union of disjoint intervals of real numbers (only for suitable variable types).

Differential Revision: [D51816908](https://our.internmc.facebook.com/intern/diff/D51816908/)

manuelcandales committed:
[InputGen] introduce variable space

This diff introduces the VariableSpace class.

A VariableSpace needs to be initialized with a variable type. Its purpose is to describe a space of values for that variable type. This space can be discrete (like a Python set) or it can be a union of disjoint intervals of real numbers (only for suitable variable types).

Differential Revision: [D51816908](https://our.internmc.facebook.com/intern/diff/D51816908/)

[ghstack-poisoned]

manuelcandales committed:
Update on "[InputGen] introduce variable space"


This diff introduces the VariableSpace class.

A VariableSpace needs to be initialized with a variable type. Its purpose is to describe a space of values for that variable type. This space can be discrete (like a Python set) or it can be a union of disjoint intervals of real numbers (only for suitable variable types).

Differential Revision: [D51816908](https://our.internmc.facebook.com/intern/diff/D51816908/)

[ghstack-poisoned]

larryliu0820 commented on torch/testing/_internal/inputgen/variable/space.py:
```diff
@@ -0,0 +1,233 @@
+
+
+class Discrete:
+    def __init__(self, values: Optional[List[Any]] = None):
+        if values is None:
+            self.initialized = False
```
IMO we shouldn't accept optional values, instead we should check if `values` has value, if not fail the constructor. I would prefer us to fail early, instead of fail at those API calls.

manuelcandales replied:
It won't fail if values is None. I actually need values to be optional.
Discrete() = Discrete(None) which is different than Discrete([])
Discrete(None) is an un-initialized discrete set
Discrete([]) is an empty set

larryliu0820 commented on torch/testing/_internal/inputgen/variable/space.py:
```diff
@@ -0,0 +1,233 @@
+        )
+
+
+class Intervals:
```
Can this be a subclass of `Discrete`? A lot of the APIs look the same

manuelcandales replied:
No. They are different objects.

larryliu0820 commented on torch/testing/_internal/inputgen/variable/space.py:
```diff
@@ -0,0 +1,233 @@
+                return
+
+    def set_lower(self, lower: float, lower_open: bool = False) -> None:
+        for ix, r in enumerate(self.intervals):
```
This seems like a O(N) solution. If you keep your interval sorted it's going to be O(logN)

manuelcandales replied:
They are sorted. You are right that I could do binary search here. However, for the present use case, the Intervals objects will contain at most 2 or 3 intervals. I don't even know if binary search would be more efficient with so little members, and has the downside of potentially introducing bugs.
True, for any generic Intervals object this is inefficient. But IMO there is no practical need for complicating things at this stage.

manuelcandales committed:
Update on "[InputGen] introduce variable space"


This diff introduces the VariableSpace class.

A VariableSpace needs to be initialized with a variable type. Its purpose is to describe a space of values for that variable type. This space can be discrete (like a Python set) or it can be a union of disjoint intervals of real numbers (only for suitable variable types).

Differential Revision: [D51816908](https://our.internmc.facebook.com/intern/diff/D51816908/)

[ghstack-poisoned]

manuelcandales committed:
Update on "[InputGen] introduce variable space"


This diff introduces the VariableSpace class.

A VariableSpace needs to be initialized with a variable type. Its purpose is to describe a space of values for that variable type. This space can be discrete (like a Python set) or it can be a union of disjoint intervals of real numbers (only for suitable variable types).

Differential Revision: [D51816908](https://our.internmc.facebook.com/intern/diff/D51816908/)

[ghstack-poisoned]

manuelcandales committed:
Update on "[InputGen] introduce variable space"


This diff introduces the VariableSpace class.

A VariableSpace needs to be initialized with a variable type. Its purpose is to describe a space of values for that variable type. This space can be discrete (like a Python set) or it can be a union of disjoint intervals of real numbers (only for suitable variable types).

Differential Revision: [D51816908](https://our.internmc.facebook.com/intern/diff/D51816908/)

[ghstack-poisoned]

manuelcandales committed:
Update on "[InputGen] introduce variable space"


This diff introduces the VariableSpace class.

A VariableSpace needs to be initialized with a variable type. Its purpose is to describe a space of values for that variable type. This space can be discrete (like a Python set) or it can be a union of disjoint intervals of real numbers (only for suitable variable types).

Differential Revision: [D51816908](https://our.internmc.facebook.com/intern/diff/D51816908/)

[ghstack-poisoned]

manuelcandales committed:
Update on "[InputGen] introduce variable space"


This diff introduces the VariableSpace class.

A VariableSpace needs to be initialized with a variable type. Its purpose is to describe a space of values for that variable type. This space can be discrete (like a Python set) or it can be a union of disjoint intervals of real numbers (only for suitable variable types).

Differential Revision: [D51816908](https://our.internmc.facebook.com/intern/diff/D51816908/)

[ghstack-poisoned]
